### DISCUSSION TITLE
Server Side Rendering

### DISCUSSION THREAD

**Discussion by iacore** [rust-lang/crates.io#4404] [Feature Requests]
Currently, crates.io need Javascript to work. It doesn't need to be. I want to add server-side rendering to crates.io to make the page load faster. If the API calls are internal, the page will load much faster.

Also, on https://crates.io/crates/clap, the website fetch from https://play.rust-lang.org/meta/crates. It makes no sense. There is a lot to optimize.

**Comment by iacore** [comment]
The code base is using [fastboot](https://ember-fastboot.com/quickstart), but it's not doing anything.

(search for @service fastboot)

**Comment by Turbo87** [comment]
we have experimented with using fastboot in the past, but back then it caused us to use up a lot more memory on the servers and with the constrained resources we have available it wasn't feasible yet. I think we would have to separately server frontend and backend from different machines if we want to go down this route, to not worsen the performance of the API, which is also used by things like cargo and other services.

> on https://crates.io/crates/clap, the website fetch from https://play.rust-lang.org/meta/crates.

that is intentional. on the crates that are available on the playground we show a "try on the playground" button, but to be able to do that we need to know what crates are available there. we only load that data once on app startup though and not on every crate page.

**Reply by iacore** [reply]
I want to try it again. In the current codebase `index.html` hardcoded `assets/cargo.js`. From my limited understanding of ember, that's bad practice.

**Reply by iacore** [reply]
I feel like ember is the culprit. I'll try to make a demo using svelte.

**Reply by iacore** [reply]
ember's compiler is garbage too. When I run `yarn build`, the compiler uses a lot of CPU%, and even after the command finishes, there is still one thread lingering, using 100% CPU. I don't understand this.

**Reply by Turbo87** [reply]
> From my limited understanding of ember, that's bad practice.

I don't know why you think that that is bad practice?

> I feel like ember is the culprit. I'll try to make a demo using svelte.

feel free to work on that, but a large scale rewrite like that is very unlikely to be merged any time soon for various reasons, including our current code review capacity. I'm also slightly sceptical whether it would solve the issue, since we would still need to run Node.js on the servers, which creates the additional memory pressure.

**Reply by iacore** [reply]
> I don't know why you think that that is bad practice?
It's hard for other tools to do static analsis.

**Comment by carols10cents** [comment]
> ember's compiler is garbage too.

We don't bash tools like that here. Please try to be more constructive in
the future.

**Reply by iacore** [reply]
Sorry, I only can give 2GB of memory to Firefox, a code editor, and ember. If I use VS Code, then my computer would freeze when I try to run `ember serve`. Then I have to wait till Linux OOM kills it or restart manually.

**Comment by iacore** [comment]
So, I recreated the index page of crates.io minus CSS and SVG icons using SvelteKit. The page will load without Javascript.

I think the code is cleaner than the official crates.io due to the use of modules, so possibly easier to do code review?.

We still need node or another Javascript engine due to Svelte compiling template to javascript (es5 or commonjs) module.

Code: https://github.com/locriacyber/crates.io-demo

![Rendered page](https://user-images.githubusercontent.com/74560659/148092917-f3ca70d5-fba6-4e09-8cbb-741db9b50e7a.png)

**Reply by locks** [reply]
Are you familiar with https://lib.rs/ ? It might be what you are looking for.

**Reply by iacore** [reply]
No. Thank you! How did you know about the website?

**Reply by AlexCzar** [reply]
Thanks, came here while searching if a similar issue was already filed and found an alternative solution, looks like lib.rs is a good replacement for crates.io, at least for my use case: quick command line lookup :)

**Comment by apiraino** [comment]
Now that crates.io is moving to Svelte, are there any chances for a minimal working version that doesn't need javascript?

Currently both the emberjs and svelte versions of crates.io show a black page when visiting without javascript activated.

**Reply by iacore** [reply]
Does lib.rs work for you?

**Reply by apiraino** [reply]
I've been using lib.rs since years :-)

But I'd still prefer if the official crates index was usable without JavaScript (all links point to crates.io so it's always slightly inconvenient)
